videos/youtube.py

```diff
@@ -0,0 +1,276 @@
+    """
+    response = None
+    error = None
+
+    retry = 0
+    retry_exceptions = (OSError, http.client.HTTPException)
+    retry_statuses = [500, 502, 503, 504]
+
+    while response is None:
+        try:
+            _, response = request.next_chunk()
+            if response is not None and "id" not in response:
+                raise YouTubeUploadException("YouTube upload failed: %s" % response)
+        except HttpError as e:
+            if e.resp.status in retry_statuses:
+                error = e
+            else:
+                raise
+        except retry_exceptions as e:
+            error = e
+
+        if error is not None:
+            retry += 1
+            if retry > max_retries:
+                log.error("Final upload failure")
+                raise YouTubeUploadException(
+                    "Retried YouTube upload 10x, giving up"
+                ) from error
+            sleep_time = 2 ** retry
+            time.sleep(sleep_time)
+
+    return response
+
+
+def strip_bad_chars(txt):
+    """
+    Remove any characters forbidden by Youtube (<, >) from text
+
+    Args:
+        txt(str): Text to remove characters from.
+
+    Returns:
+        str: Text without bad characters
+    """
+    return re.sub("<|>", "", txt)
+
+
+class YouTubeApi:
+    """
+    Class interface to YouTube API calls
+    """
+
+    client = None
+    s3 = None
+
+    def __init__(self):
+        """
+        Generate an authorized YouTube API client and S3 client
+        """
+        credentials = oauth2client.client.GoogleCredentials(
+            settings.YT_ACCESS_TOKEN,
+            settings.YT_CLIENT_ID,
+            settings.YT_CLIENT_SECRET,
+            settings.YT_REFRESH_TOKEN,
+            None,
+            "https://accounts.google.com/o/oauth2/token",
+            None,
+        )
+        authorization = credentials.authorize(httplib2.Http())
+        credentials.refresh(authorization)
+        self.client = build("youtube", "v3", credentials=credentials)
+        self.s3 = boto3.client("s3")
+
+    def video_status(self, video_id):
+        """
+        Checks the  status of a video. 'processed' = ready for viewing.
+
+        Args:
+            video_id(str): YouTube video id
+
+        Returns:
+            str: status of the YouTube video
+
+        """
+        results = self.client.videos().list(part="status", id=video_id).execute()
+        return results["items"][0]["status"]["uploadStatus"]
+
+    def list_captions(self, video_id):
+        """
+        List the captions available for a YouTube video
+
+        Args:
+            video_id(str): YouTube video id
+
+        Returns:
+            dict: List of captions in JSON format
+        """
+        results = (
+            self.client.captions().list(part="snippet", videoId=video_id).execute()
+        )
+
+        return {item["snippet"]["language"]: item["id"] for item in results["items"]}
+
+    def upload_caption(self, caption, video_id):
```

Comment:
Does this come from the ol-video-services codebase? Doesn't look like it's currently used in this PR but makes sense to upload to youtube when we get a new caption file from 3play

Reply:
Thanks for catching this, the functions were copied from OVS but I think it's likely they will need to be tweaked once 3play captions are available, so I'm going to cut them out, then add them back in a future PR once they're needed.